### ISSUE TITLE
getOwnerPage() fails when... SiteTree ClassName is not a table / Page has no additional data columns than it's ancestor 

### ISSUE THREAD

**Issue by rafaeldsousa** [silverstripe/silverstripe-elemental#946]
- [This commit](https://github.com/silverstripe/silverstripe-elemental/commit/a0a4ea06d2f32c2a0778142f27891ab583d1c6cf) introduces a new filter SiteTree.ClassName when getting the OwnerPage of an ElementalArea
- We have `CustomListPage` (with no `$db` fields) that extends `GenericContentPage`. So page records are recorded on `GenericContentPage` table
- But on `SiteTree` the **ClassName** of the page is saved as `CustomListPage` not `GenericContentPage`
-  With the above comment in mind, when looking for the OwnerPage and with the new ClassName filter, the page isn't found, since the ClassName coming to Elemental is `GenericContentPage`  but `SiteTree` has the page class name as `CustomListPage`

What we could identify is that Elemental is assuming that all classes have corresponding tables, but It breaks on classes that don't have db columns (e.g. classes using their parent class table) . In these cases the class recorded on the SiteTree does not exist as a table.

**Comment by rafaeldsousa** [comment]
@michalkleiner I forgot to mention that the above results in the elemental areas of the affected pages not to appear on the live version, when the page is in Draft mode. That's the biggest problem we're facing with this. 

From the example above
- Edit the page/elements on page `CustomListPage`  and just save (leaving it in Draft mode (Stage version)
- Latest published version of elements (in elemental area) are expected to appear in FE
- Actual result is that no elements from elemental area are displayed in FE (until page is published once again)

Our workaround in the mean time was to create a **tempField** `$db` field for `CustomListPage` and create a build task to populate the new `CustomListPage` related tables.

**Comment by kinglozzer** [comment]
@rafaeldsousa What’s the value of the `OwnerClassName` field? The `$class` variable in the filter `"\"{$baseTable}\".\"ClassName\" = ?" => $class` comes from that field.

If you’re editing a `CustomListPage`, as far as I can tell it should be set to `CustomListPage` but in your example it looks like it’s set to `GenericContentPage` instead. Is the problem possibly that we’re storing the wrong value for `OwnerClassName`?

**Comment by rafaeldsousa** [comment]
@kinglozzer `$class` is being picked up as `GenericContentPage`. At some point, which I haven't debugged that far, there's an assumption that the page class is always the same as of it's owner (or table where it's saved). In the example above, `CustomListPage` has no `$db` fields and extends `GenericContentPage`, so no table is created for `CustomListPage` and all page records are saved on `GenericContentPage`, but on `SiteTree` the same page record has the `ClassName` as `CustomListPage`, the filter being done in the commit mentioned above also assumes that the className from `baseTable` (SiteTree in this case) is the same as the class/table name of the owner.

- **Y** has `$db` fields
- **X** extends **Y**
- **X** has no fields
- **X** records saved on **Y** table
- SiteTree **X** records ClassName saved **X**
- **X** `OwnerClassName` on `ElementalArea.php` return **Y**
- `$baseTable` in new filter is **SiteTree**
- `$class` is **Y**
- No records are found due to ClassName and AreaID not finding any matches

**Comment by michalkleiner** [comment]
@rafaeldsousa the commit introducing the check might be a red herring here.
Can you check what do you have in the `OwnerClassName` column in the `ElementalArea` table for the page instance of the page type without extra db fields? It's possible the mechanism is not storing the right class to the db in the first place.

**Comment by rafaeldsousa** [comment]
@michalkleiner it seems like it's just as you've said, the className saved on `ElementalArea` is not the correct one.
<img width="891" alt="Screen Shot 2021-11-23 at 12 29 42 PM" src="https://user-images.githubusercontent.com/39249777/142950242-8dd5e7f4-4969-451b-8f5c-72d013816817.png">

Worth noting that after I've applied the workaround, I've mentioned above (placeholder db on the pagetype that didn't have fields and build task to publish pages) the `OwnerClassName` on `ElementalArea` got updated to `CustomListPage`

**Comment by kinglozzer** [comment]
So it sounds like we need to debug why `OwnerClassName` is set to the nearest ancestor with its own table instead of being set to the correct “instance” class name?

**Comment by michalkleiner** [comment]
The problem relates to `ElementalArea->supportedPageTypes` which only considers classes backed by db tables. It iterates over the list of such classes and tries to find all elemental area relations. When it finds one that corresponds with the current page, it assigns the OwnerClassName. But because it's never in the initial list, it will never be assigned correctly.

We can try changing `ClassInfo::getValidSubClasses(DataObject::class)` in the `supportedPageTypes` method to `ClassInfo::getValidSubClasses(DataObject::class, true)`, to include unbacked classes as well, and see what happens.

**Comment by rafaeldsousa** [comment]
@michalkleiner Yep that's where I thought it would be coming from, I wasn't looking for where `OwnerClassName` gets set and thought might be there. But looking at the `getValidSubClasses` even if we add the `true` flag to unbacked classes, this is what happens
```
if ($includeUnbacked) {
    $table = DataObject::getSchema()->tableName($class);
    $classes = DB::get_schema()->enumValuesForField($table, 'ClassName');
} else {
    $classes = static::subclassesFor($class);
}
```

Just looks like it still relies on DB table names and the `CustomListPage` won't have a table in this example, wouldn't that fail still ?
